File: go/src/util/BoardJob.go
package util

import "strconv"

/**
 * Represents a unit of work that can be worked upon.
 */
type BoardJob struct {
    PegsLeft int
    InitialPeg int
    TotalMoves int
    MoveString string
    Board []bool
}

/**
 * Creates the next board job given the previous one and the new state of the board.
 */
func NewBoardJobWithMove(job BoardJob, newBoard []bool, moveString string) BoardJob {
    pegsLeft := job.PegsLeft
    initialPeg := job.InitialPeg
    totalMoves := job.TotalMoves + 1
    newMoveString := job.MoveString + "\n" + moveString
    return BoardJob{pegsLeft, initialPeg, totalMoves, newMoveString, newBoard}
}

/**
 * Returns the string representation of the board job.
 */
func BoardJobToString(b BoardJob) string {
    return "(" + strconv.Itoa(b.InitialPeg) + "," + strconv.Itoa(b.TotalMoves) + ")" + b.MoveString
}

File: java/BoardJob.java
/**
 * Represents a "job" that needs to be done on a thread.
 */
public class BoardJob implements Comparable {
    public int pegsLeft;
    public int initialPeg;
    public int totalMoves;
    public boolean[] board;
    public String moveString;

    public BoardJob(int pegsLeft, int initialPeg, int totalMoves, boolean[] board, String moveString) {
        this.pegsLeft = pegsLeft;
        this.initialPeg = initialPeg;
        this.totalMoves = totalMoves;
        this.board = board;
        this.moveString = moveString;
    }

    @Override
    public int compareTo(Object o) {
        if (this.pegsLeft > ((BoardJob) o).pegsLeft) {
            return -1;
        } else if (this.pegsLeft < ((BoardJob) o).pegsLeft){
            return 1;
        }
        if (this.initialPeg > ((BoardJob) o).initialPeg) {
            return -1;
        } else if (this.initialPeg < ((BoardJob) o).initialPeg){
            return 1;
        }

        return 0;
    }

    @Override
    public String toString() {
        return "(" + initialPeg + "," + totalMoves + ")" + moveString;
    }
}
